packages/valory/skills/dynamic_nft_abci/models.py

```diff
@@ -70,5 +117,176 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
+        self.api = pygsheets.authorize(service_file=service_auth_file_path)
+        self.leaderboard_sheet_id = self.ensure("leaderboard_sheet_id", kwargs)
+        self.leaderboard_sheet_name = self.ensure("leaderboard_sheet_name", kwargs)
+        # The following line accesses Google service to open the spreadsheet.
+        # During e2e tests, this is not possible since the service auth data
+        # is dummy. For this reason, we do not execute if we detect we are in an e2e test.
+        self.sheet = (
+            self.api.open_by_key(self.leaderboard_sheet_id).worksheet(
+                "title", self.leaderboard_sheet_name
+            )
+            if "my_dummy_project_id" not in service_auth
+            else None
+        )
+        self.name_col = self.ensure("name_col", kwargs)
+        self.address_col = self.ensure("address_col", kwargs)
+        self.points_col = self.ensure("points_col", kwargs)
+        self.discord_id_col = self.ensure("discord_id_col", kwargs)
+        self.observation_interval = self.ensure("observation_interval", kwargs)
+
+        self.lock = Lock()
+        manager = Manager()
+        self.linking_wallets = manager.dict(lock=True)
+
+        super().__init__(*args, **kwargs)
+
+    def ensure(self, keyword: str, kwargs: Dict) -> Any:
+        """Ensure a keyword argument."""
+        value = kwargs.pop(keyword, None)
+        if value is None:
+            raise ValueError(
+                f"Value for {keyword} is required by {self.__class__.__name__}."
+            )
+        return value
+
+    def create(self, discord_id, name, address=DEFAULT_ADDRESS):
+        """Create a new user."""
+        rows = self.read()
+        last_entry_index = len(rows) + 1 if rows else 1  # Take the header into account
+
+        # Build the values in the correct order: column indices are configurable
+        index_to_value = {
+            self.discord_id_col: discord_id,
+            self.address_col: address,
+            self.name_col: name,
+            self.points_col: DEFAULT_POINTS,
+        }
+        ordered_values = [index_to_value[k] for k in sorted(index_to_value.keys())]
+
+        # Insert the new row after last_entry_index
+        self.sheet.insert_rows(row=last_entry_index, number=1, values=ordered_values)
+
+    def row_to_dict(self, row):
+        """Returns the data for a specific formatted for the HTTP response"""
+        return {
+            "name": row[self.name_col - 1],
+            "address": row[self.address_col - 1],
+            "points": row[self.points_col - 1],
+            "discord_id": row[self.discord_id_col - 1],
+        }
+
+    @timed_lru_cache(seconds=CACHE_EXPIRATION_SECONDS, maxsize=CACHE_MAXSIZE)
+    def _read_sheet(self):
+        """Reads the sheet and returns the sheet as a matrix"""
+        if self.lock.locked():
+            return self.rows
+        with self.lock:
+            self.rows = self.sheet.get_all_values(returnas="matrix")
+        return self.rows
```

Comment:
This pattern here is not in keeping with AEA architecture. You end up making a network request from the skill.